src/vos/vos_pool.c

```diff
@@ -96,6 +96,134 @@ vos_meta_readv(struct umem_store *store, struct umem_store_iod *iod, d_sg_list_t
 	return vos_meta_rwv(store, iod, sgl, false);
 }
 
+
+#define META_READ_BATCH_SIZE (1024 * 1024)
+
+struct meta_load_control {
+	ABT_mutex		mlc_lock;
+	ABT_cond		mlc_cond;
+	int			mlc_rc;
+	int			mlc_inflights;
+	int			mlc_wait_finished;
+};
+
+struct meta_load_arg {
+	daos_off_t		  mla_off;
+	uint64_t		  mla_read_size;
+	char		         *mla_start;
+	struct umem_store        *mla_store;
+	struct meta_load_control *mla_control;
+};
+
+#define META_READ_QD_NR	4
+
+static inline void
+vos_meta_load_fn(void *arg)
+{
+	struct meta_load_arg	  *mla = arg;
+	struct umem_store_iod	  iod;
+	d_sg_list_t		  sgl;
+	d_iov_t			  iov;
+	int			  rc;
+	struct meta_load_control *mlc = mla->mla_control;
+
+	D_ASSERT(mla != NULL);
+	iod.io_nr             = 1;
+	iod.io_regions        = &iod.io_region;
+	iod.io_region.sr_addr = mla->mla_off;
+	iod.io_region.sr_size = mla->mla_read_size;
+
+	sgl.sg_iovs   = &iov;
+	sgl.sg_nr     = 1;
+	sgl.sg_nr_out = 0;
+	d_iov_set(&iov, mla->mla_start, mla->mla_read_size);
+
+	rc = vos_meta_rwv(mla->mla_store, &iod, &sgl, false);
+	if (!mlc->mlc_rc && rc)
+		mlc->mlc_rc = rc;
+
+	D_FREE(mla);
+	mlc->mlc_inflights--;
```

Comment:
For completeness is it good to protect updating this using associated mutex lock? This may not be a requirement given that all contending ULTs are running on the same thread.

Reply:
I think it fine, we use one thread here.